### PR TITLE
[Transformations] Leftovers: FuseU4WeightsAndZeroPoint transformation

### PR DESCRIPTION
**Pull request by v-Golubev**
Main PR: #20503

### CODE REVIEW

File: src/common/transformations/src/transformations/common_optimizations/convert_u4_weights_zero_point_to_scalar.cpp [hunk 1]
```diff
@@ -51,21 +51,19 @@ ov::pass::ConvertU4WeightsZeroPointToScalar::ConvertU4WeightsZeroPointToScalar()
...
             return false;
 
-        zero_point_shape.insert(zero_point_shape.begin(), weights_rank - zero_point_rank, 1);
-        for (size_t i = 0; i < weights_rank; ++i) {
-            if (zero_point_shape[i] > weights_shape[i])
+        for (size_t i = 0; i < zero_point_shape.size(); ++i) {
+            if (*(zero_point_shape.rbegin() + i) > *(weights_shape.rbegin() + i)) {
                 return false;
+            }
         }
```
Comment (praasz) [inline src/common/transformations/src/transformations/common_optimizations/convert_u4_weights_zero_point_to_scalar.cpp:67]:
```cpp
if(weights_shape.size() < zero_point_shape.size() 
   || !std::equal(zero_point_shape.rbegin(), zero_point_shape.rend(), weights_shape.rbegin(), std::less_equal<size_t>())){
    return false;
}
```
Consider change `weights` and `zero_point` shapes check (it should be same as explicit loop here.

File: src/common/transformations/include/transformations/utils/utils.hpp [hunk 2]
```diff
@@ -196,6 +196,10 @@ TRANSFORMATIONS_API void visit_shape_path(ov::Node* node,
                                           std::unordered_set<ov::Node*>& visited,
                                           std::function<void(ov::Node*)> func);
 
+TRANSFORMATIONS_API void visit_constant_path(ov::Node* node,
```
Comment (praasz) [inline src/common/transformations/include/transformations/utils/utils.hpp:199]: Consider add doxygen to explain what is doing on constant path?

File: src/common/transformations/src/transformations/utils/utils.cpp [hunk 3]
```diff
@@ -266,6 +266,28 @@ void visit_shape_path(Node* node, std::unordered_set<ov::Node*>& visited, std::f
     }
 }
 
+void visit_constant_path(ov::Node* node, std::unordered_set<ov::Node*>& visited, std::function<void(ov::Node*)> func) {
```
Comment (praasz) [inline src/common/transformations/src/transformations/utils/utils.cpp:269]: The function body looks very similar to `visit_shape_path`, maybe is possible to add some internal helper which can implement both functions?
The difference is in continue or assert which could passed as some parameter (predicate)?

File: src/common/transformations/src/transformations/utils/utils.cpp [hunk 4]
```diff
@@ -266,6 +266,28 @@ void visit_shape_path(Node* node, std::unordered_set<ov::Node*>& visited, std::f
...
+    std::deque<ov::Node*> nodes{node};
+    while (!nodes.empty()) {
+        auto curr_node = nodes.front();
+        nodes.pop_front();
+        OPENVINO_ASSERT(!ov::is_type<opset1::Parameter>(curr_node),
+                        "visit_constant_path is called for non-constant path.");
+
+        func(curr_node);
+        for (auto& input_value : curr_node->input_values()) {
+            const auto input_node = input_value.get_node();
```
Comment (praasz) [inline src/common/transformations/src/transformations/utils/utils.cpp:282]:
```suggestion
            const auto& input_node = input_value.get_node();
```
Or copy is intentional and requried?

**Review by praasz — APPROVED** [review]

File: src/common/transformations/src/transformations/common_optimizations/convert_u4_weights_zero_point_to_scalar.cpp (hunk 1 above)
Comment (v-Golubev) [reply src/common/transformations/src/transformations/common_optimizations/convert_u4_weights_zero_point_to_scalar.cpp]: Thanks for the suggestion, changed

File: src/common/transformations/include/transformations/utils/utils.hpp (hunk 2 above)
Comment (v-Golubev) [reply src/common/transformations/include/transformations/utils/utils.hpp:215]: Added, thanks

File: src/common/transformations/src/transformations/utils/utils.cpp (hunk 3 above)
Comment (v-Golubev) [reply src/common/transformations/src/transformations/utils/utils.cpp:279]: Good idea. The internal helper reused in both functions

**Review by jane-intel — APPROVED** [review]

**Pull request merged**